src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,386 @@
+        }
+
+        private void ConvertToBool(TypeSymbol type, BinaryExpressionSyntax node, BinaryOperatorKind binaryOperator, DiagnosticBag diagnostics,
+            out Conversion boolConversion, out UnaryOperatorSignature boolOperator)
+        {
+            BoundExpression comparisonResult = new BoundTupleOperandPlaceholder(node, type);
+            var boolean = GetSpecialType(SpecialType.System_Boolean, diagnostics, node);
+
+            // Is the operand implicitly convertible to bool?
+
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            var conversion = this.Conversions.ClassifyConversionFromExpression(comparisonResult, boolean, ref useSiteDiagnostics);
+            diagnostics.Add(node, useSiteDiagnostics);
+
+            if (conversion.IsImplicit)
+            {
+                boolConversion = conversion;
+                boolOperator = default;
+                return;
+            }
+
+            // It was not. Does it implement operator true (or false)?
+
+            UnaryOperatorKind boolOpKind;
+            switch (binaryOperator)
+            {
+                case BinaryOperatorKind.Equal:
+                    boolOpKind = UnaryOperatorKind.False;
+                    break;
+                case BinaryOperatorKind.NotEqual:
+                    boolOpKind = UnaryOperatorKind.True;
+                    break;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(binaryOperator);
+            }
+
+            LookupResultKind resultKind;
+            ImmutableArray<MethodSymbol> originalUserDefinedOperators;
+            var best = this.UnaryOperatorOverloadResolution(boolOpKind, comparisonResult, node, diagnostics, out resultKind, out originalUserDefinedOperators);
+            if (best.HasValue)
+            {
+                boolConversion = best.Conversion;
+                boolOperator = best.Signature;
+                return;
+            }
+
+            // It did not. Give a "not convertible to bool" error.
+
+            GenerateImplicitConversionError(diagnostics, node, conversion, comparisonResult, boolean);
+            boolConversion = Conversion.NoConversion;
+            boolOperator = default;
+            return;
+        }
+
+        private TupleBinaryOperatorInfo BindTupleDynamicBinaryOperatorSingleInfo(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            // This method binds binary == and != operators where one or both of the operands are dynamic.
+            Debug.Assert((object)left.Type != null && left.Type.IsDynamic() || (object)right.Type != null && right.Type.IsDynamic());
+
+            bool hasError = false;
+            bool leftValidOperand = IsLegalDynamicOperand(left);
+            bool rightValidOperand = IsLegalDynamicOperand(right);
+
+            if (!leftValidOperand || !rightValidOperand)
```

Comment:
Consider inlining the two locals - the locals are only used here and it would avoid evaluating the right if the left is false.

Reply:
Actually, it looks like we have this same check (and error message) in several `Binder` methods. Consider extracting a helper method.

Reply:
It would extract cleanly here, but not in `BindDynamicBinaryOperator` (because of local `leftValidOperand` which is used later on).

---
In reply to: [167709296](https://github.com/dotnet/roslyn/pull/23356#discussion_r167709296) [](ancestors = 167709296)